### ISSUE TITLE
GPU memory usage keeps increasing till OOM when qlora tuning with 1 Arc 770 card

### ISSUE THREAD

**Issue by kyang-06** [intel-analytics/BigDL#9675]
Hi team,

I encountered this issue by running [qalora_finetune_llama2_7b_arc_1_card.sh](https://github.com/intel-analytics/BigDL/blob/main/python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/qalora_finetune_llama2_7b_arc_1_card.sh)

When qlora tuning llama-7b, GPU memory usage starts from ~5G to more than 10G (when epoch=0.32), and eventually OOM.
The only one modification that I made is replacing `load_dataset` with `load_from_disk`, and still alpaca dataset

Can you give me some hints on what happened?

**Comment by rnwang04** [comment]
Hi @kyang-06 ,
When running qlora / qalora, GPU memory will increase at the beginning, it's normal. But I think it won't continuously increase to out of GPU memory，and I don't observe this phenomenon before.

Do you make any other changes to the training scirpt (except `load_from _disk`) or do you set any other env variables?
By the way, have you ever tried qlora(https://github.com/intel-analytics/BigDL/blob/main/python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/finetune_llama2_7b_arc_1_card.sh) instead of qalora ? Will it OOM too ?

**Comment by cyita** [comment]
### Test result on arc02
It seems both qlora and qalora have memory leaks on arc.
Llama-7b-chat-hf, same params as [qalora_finetune_llama2_7b_arc_1_card.sh](https://github.com/intel-analytics/BigDL/blob/main/python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/qalora_finetune_llama2_7b_arc_1_card.sh)

#### QALoRA, 0.34 epoch, memory usage
12:17, 4942MB

12:19, 9486MB    `+ 2 min`, `+ 4544MB`

12:23, 9825MB    `+ 4 min`, `+ 339MB`

13:35, 10350MB  `+ 72 min`, `+525MB`

#### QLoRA, 0.32 epoch, memory usage
14:06, 4980MB

14:07, 11283MB   `+ 1 min`, `+6303MB`

14:12, 11730MB   `+ 5 min`, `+ 447MB`

15:22, 12284MB   `+ 70 min`, `+ 554MB`

**Comment by cyita** [comment]
> ### Test result on arc02
> It seems both qlora and qalora have memory leaks on arc. `Maybe related to eval(?)` Llama-7b-chat-hf, same params as [qalora_finetune_llama2_7b_arc_1_card.sh](https://github.com/intel-analytics/BigDL/blob/main/python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/qalora_finetune_llama2_7b_arc_1_card.sh)
> 
> #### QALoRA, 0.34 epoch, memory usage
> 12:17, 4942MB
> 
> 12:19, 9486MB `+ 2 min`, `+ 4544MB`
> 
> 12:23, 9825MB `+ 4 min`, `+ 339MB`
> 
> 13:35, 10350MB `+ 72 min`, `+525MB`
> 
> #### QLoRA, 0.32 epoch, memory usage
> 14:06, 4980MB
> 
> 14:07, 11283MB `+ 1 min`, `+6303MB`
> 
> 14:12, 11730MB `+ 5 min`, `+ 447MB`
> 
> 15:22, 12284MB `+ 70 min`, `+ 554MB`

The `+70 min` is very close to the eval time.

**Comment by cyita** [comment]
The GPU memory will increase when first evaluation (~78min in the above case), but won't increase for the following evals(>=2).

**Comment by rnwang04** [comment]
Hi @kyang-06, could you please paste your detailed error message ?  Are you sure it is GPU OOM not CPU ?

**Comment by kyang-06** [comment]
As I tried yesterday, [https://github.com/intel-analytics/BigDL/blob/main/python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/finetune_llama2_7b_arc_1_card.sh](https://github.com/intel-analytics/BigDL/blob/main/python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/finetune_llama2_7b_arc_1_card.sh) also caused out of GPU memory. It consumes 12.5G memory at epoch=0.3, and was killed by OOM at epoch=1.2. In comparison with this, qlora finetuning halted much earlier.

I modified the shell to
`python ./alpaca_qlora_finetuning_customized.py 
    --base_model "/home/kyyx/.cache/huggingface/hub/models--meta-llama--Llama2-2-7b-chat-hf/snapshots/download" 
    --data_path "/home/kyyx/.cache/huggingface/local_datasets/tatsu_lab__alpaca" 
    --output_dir "./bigdl-qlora-alpaca"`
Just  noticed that I use inconsistent dataset with your default setting. Don't know if it has influence.

Here are some env params that I suppose may have relevance

> ACL_BOARD_VENDOR_PATH : /opt/Intel/OpenCLFPGA/oneAPI/Boards
ADVISOR_2024_DIR : /opt/intel/oneapi/advisor/2024.0
APM : /opt/intel/oneapi/advisor/2024.0/perfmodels
CCL_CONFIGURATION : cpu_gpu_dpcpp
CCL_ROOT : /opt/intel/oneapi/ccl/2021.11
CLASSPATH : /opt/intel/oneapi/mpi/2021.11/share/java/mpi.jar
CMAKE_PREFIX_PATH : /opt/intel/oneapi/tbb/2021.11/env/..:/opt/intel/oneapi/mkl/2024.0/lib/cmake:/opt/intel/oneapi/ipp/2021.10/lib/cmake/ipp:/opt/intel/oneapi/dpl/2022.3/lib/cmake/oneDPL:/opt/intel/oneapi/dnnl/2024.0/lib/cmake:/opt/intel/oneapi/dal/2024.0:/opt/intel/oneapi/compiler/2024.0
CMPLR_ROOT : /opt/intel/oneapi/compiler/2024.0
CONDA_EXE : /home/kyyx/anaconda3/bin/conda
CONDA_PREFIX : /home/kyyx/anaconda3/envs/intel-arc-py39
CONDA_PREFIX_1 : /home/kyyx/anaconda3
CONDA_PREFIX_2 : /home/kyyx/anaconda3/envs/intel-arc-py39
CONDA_PREFIX_3 : /home/kyyx/anaconda3/envs/py37
CONDA_PREFIX_4 : /home/kyyx/anaconda3/envs/py39
CONDA_PROMPT_MODIFIER : (intel-arc-py39)
CONDA_PYTHON_EXE : /home/kyyx/anaconda3/bin/python
CPATH : /opt/intel/oneapi/tbb/2021.11/env/../include:/opt/intel/oneapi/mpi/2021.11/include:/opt/intel/oneapi/mkl/2024.0/include:/opt/intel/oneapi/ippcp/2021.9/include:/opt/intel/oneapi/ipp/2021.10/include:/opt/intel/oneapi/dpl/2022.3/include:/opt/intel/oneapi/dpcpp-ct/2024.0/include:/opt/intel/oneapi/dnnl/2024.0/include:/opt/intel/oneapi/dev-utilities/2024.0/include:/opt/intel/oneapi/dal/2024.0/include/dal:/opt/intel/oneapi/compiler/2024.0/opt/oclfpga/include:/opt/intel/oneapi/ccl/2021.11/include
DALROOT : /opt/intel/oneapi/dal/2024.0
DAL_MAJOR_BINARY : 2
DAL_MINOR_BINARY : 0
DIAGUTIL_PATH : /opt/intel/oneapi/dpcpp-ct/2024.0/etc/dpct/sys_check/sys_check.sh:/opt/intel/oneapi/debugger/2024.0/etc/debugger/sys_check/sys_check.py:/opt/intel/oneapi/compiler/2024.0/etc/compiler/sys_check/sys_check.sh
DNNLROOT : /opt/intel/oneapi/dnnl/2024.0
DPL_ROOT : /opt/intel/oneapi/dpl/2022.3
FI_PROVIDER_PATH : /opt/intel/oneapi/mpi/2021.11/opt/mpi/libfabric/lib/prov:/usr/lib/x86_64-linux-gnu/libfabric
GDB_INFO : /opt/intel/oneapi/debugger/2024.0/share/info/
INFOPATH : /opt/intel/oneapi/debugger/2024.0/opt/debugger/lib                                                                                                                                                        [32/3497]
INTELFPGAOCLSDKROOT : /opt/intel/oneapi/compiler/2024.0/opt/oclfpga
INTEL_PYTHONHOME : /opt/intel/oneapi/debugger/2024.0/opt/debugger
IPPCP_TARGET_ARCH : intel64
IPPCRYPTOROOT : /opt/intel/oneapi/ippcp/2021.9
IPPROOT : /opt/intel/oneapi/ipp/2021.10
IPP_TARGET_ARCH : intel64
I_MPI_ROOT : /opt/intel/oneapi/mpi/2021.11
LD_LIBRARY_PATH : /opt/intel/oneapi/tbb/2021.11/env/../lib/intel64/gcc4.8:/opt/intel/oneapi/mpi/2021.11/opt/mpi/libfabric/lib:/opt/intel/oneapi/mpi/2021.11/lib:/opt/intel/oneapi/mkl/2024.0/lib:/opt/intel/oneapi/ippcp/2021.9/lib/:/opt/intel/oneapi/ipp/2021.10/lib:/opt/intel/oneapi/dpl/2022.3/lib:/opt/intel/oneapi/dnnl/2024.0/lib:/opt/intel/oneapi/debugger/2024.0/opt/debugger/lib:/opt/intel/oneapi/dal/2024.0/lib:/opt/intel/oneapi/compiler/2024.0/opt/oclfpga/host/linux64/lib:/opt/intel/oneapi/compiler/2024.0/opt/compiler/lib:/opt/intel/oneapi/compiler/2024.0/lib:/opt/intel/oneapi/ccl/2021.11/lib/
LIBRARY_PATH : /opt/intel/oneapi/tbb/2021.11/env/../lib/intel64/gcc4.8:/opt/intel/oneapi/mpi/2021.11/lib:/opt/intel/oneapi/mkl/2024.0/lib/:/opt/intel/oneapi/ippcp/2021.9/lib/:/opt/intel/oneapi/ipp/2021.10/lib:/opt/intel/oneapi/dpl/2022.3/lib:/opt/intel/oneapi/dnnl/2024.0/lib:/opt/intel/oneapi/dal/2024.0/lib:/opt/intel/oneapi/compiler/2024.0/lib:/opt/intel/oneapi/ccl/2021.11/lib/
MANPATH : /opt/intel/oneapi/mpi/2021.11/share/man:/opt/intel/oneapi/debugger/2024.0/share/man:/opt/intel/oneapi/compiler/2024.0/documentation/en/man/common:
MKLROOT : /opt/intel/oneapi/mkl/2024.0
MOTD_SHOWN : pam
NLSPATH : /opt/intel/oneapi/mkl/2024.0/share/locale/%l_%t/%N:/opt/intel/oneapi/compiler/2024.0/lib/locale/%l_%t/%N
OCL_ICD_FILENAMES : libintelocl_emu.so:libalteracl.so:/opt/intel/oneapi/compiler/2024.0/lib/libintelocl.so
ONEAPI_ROOT : /opt/intel/oneapi
PATH : /opt/intel/oneapi/vtune/2024.0/bin64:/opt/intel/oneapi/mpi/2021.11/opt/mpi/libfabric/bin:/opt/intel/oneapi/mpi/2021.11/bin:/opt/intel/oneapi/mkl/2024.0/bin:/opt/intel/oneapi/dpcpp-ct/2024.0/bin:/opt/intel/oneapi/dev-utilities/2024.0/bin:/opt/intel/oneapi/debugger/2024.0/opt/debugger/bin:/opt/intel/oneapi/compiler/2024.0/opt/oclfpga/bin:/opt/intel/oneapi/compiler/2024.0/bin:/opt/intel/oneapi/advisor/2024.0/bin64:/home/kyyx/.local/bin:/home/kyyx/.local/bin:/home/kyyx/anaconda3/envs/intel-arc-py39/bin:/home/kyyx/anaconda3/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
PKG_CONFIG_PATH : /opt/intel/oneapi/vtune/2024.0/include/pkgconfig/lib64:/opt/intel/oneapi/tbb/2021.11/env/../lib/pkgconfig:/opt/intel/oneapi/mpi/2021.11/lib/pkgconfig:/opt/intel/oneapi/mkl/2024.0/lib/pkgconfig:/opt/intel/oneapi/ippcp/2021.9/lib/pkgconfig:/opt/intel/oneapi/dpl/2022.3/lib/pkgconfig:/opt/intel/oneapi/dnnl/2024.0/lib/pkgconfig:/opt/intel/oneapi/dal/2024.0/lib/pkgconfig:/opt/intel/oneapi/compiler/2024.0/lib/pkgconfig:/opt/intel/oneapi/ccl/2021.11/lib/pkgconfig/:/opt/intel/oneapi/advisor/2024.0/include/pkgconfig/lib64:
PYTHONPATH : /opt/intel/oneapi/advisor/2024.0/pythonapi
SETVARS_COMPLETED : 1
SHLVL : 2
SSH_CONNECTION : 10.238.224.100 8427 172.16.123.101 22
TBBROOT : /opt/intel/oneapi/tbb/2021.11/env/..
VTUNE_PROFILER_2024_DIR : /opt/intel/oneapi/vtune/2024.0
VTUNE_PROFILER_DIR : /opt/intel/oneapi/vtune/2024.0
XDG_DATA_DIRS : /usr/local/share:/usr/share:/var/lib/snapd/desktop
_ : /home/kyyx/anaconda3/envs/intel-arc-py39/bin/python

**Comment by kyang-06** [comment]
Sorry for not refreshing the page, the raised error:
`my_script_ft.sh: line 21: 3034711 Killed`
There is no more python error details.
You mentioned a good point. Please let me check again if it is CPU OOM

**Comment by rnwang04** [comment]
> Sorry for not refreshing the page, the raised error: `my_script_ft.sh: line 21: 3034711 Killed` There is no more python error details. You mentioned a good point. Please let me check again if it is CPU OOM

Sure, for me, the error message looks more like CPU OOM error.
BTW, here are two tips maybe you can use to better monitor memory usage:

- `watch -t -n 1 "sudo xpu-smi stats -d 0 | grep \"GPU Memory Used\" |  tee -a gpu_memory.log"` -> can save GPU memory usage to log
- `top -d 1 -b | grep "MiB Mem" > cpu_memory.log` -> can save CPU memory usage to log

Hope it can help : )

**Comment by kyang-06** [comment]
Hi team. I can confirm that the error is caused by CPU OOM.
So is there any suggestion on solving it?

Log of when the program was killed:
run [finetune_llama2_7b_arc_1_card.sh](https://github.com/intel-analytics/BigDL/blob/main/python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/finetune_llama2_7b_arc_1_card.sh)
**GPU**
`sudo xpu-smi dump -d 0 -m 0,1,2,17,18 -i 60 -n 10000`
> Timestamp, DeviceId, GPU Utilization (%), GPU Power (W), GPU Frequency (MHz), GPU Memory Bandwidth Utilization (%), GPU Memory Used (MiB)
03:27:19.000,    0,  N/A, 44.57, 2400,  N/A, 12063.71
03:28:19.000,    0,  N/A, 44.91, 2400,  N/A, 12063.71
03:29:19.000,    0,  N/A, 44.44, 2400,  N/A, 12063.71
03:30:19.000,    0,  N/A, 44.16, 2400,  N/A, 22.46
03:31:19.000,    0,  N/A, 44.16, 2400,  N/A, 22.46
03:32:19.000,    0,  N/A, 44.15, 2400,  N/A, 22.46

When killed GPU memory still has ~4GB free.

**CPU**
(Thank you for providing example script, but it couldn't normally print log.)
```
for i in `seq 0 1000`; 
do
 date +"%Y-%m-%d %H:%M:%S" | tee -a cpu_memory.log
 free -m | grep "Mem" | tee -a cpu_memory.log
 sleep 1m
done
```

>    Timestamp            total        used        free      shared  buff/cache   available
2023-12-16 03:27:32 Mem:           31915       31253         207          35         454         255
2023-12-16 03:28:32 Mem:           31915       31253         206          35         455         254
2023-12-16 03:29:32 Mem:           31915       31260         200          35         454         248
2023-12-16 03:30:32 Mem:           31915         788       30527          29         598       30652
2023-12-16 03:31:32 Mem:           31915         792       30513          29         609       30642

**Comment by jason-dai** [comment]
> Hi team. I can confirm that the error is caused by CPU OOM. So is there any suggestion on solving it?

Yes, larger CPU memory is needed with running large number of training steps. 

A simple walkaround of this issue is to restart the training from a previous checkpoint, e.g., see `resume_from_checkpoint` parameter in https://github.com/intel-analytics/BigDL/tree/main/python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora

**Comment by plusbang** [comment]
> > Hi team. I can confirm that the error is caused by CPU OOM. So is there any suggestion on solving it?
> 
> Yes, larger CPU memory is needed with running large number of training steps.
> 
> A simple walkaround of this issue is to restart the training from a previous checkpoint

yeah, please refer to https://github.com/intel-analytics/BigDL/tree/main/python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora#4-optional-resume-training for information about how to resume training.

**Comment by kyang-06** [comment]
Got it. Thank you for the explanation!
